### PR TITLE
Measure scoring, reportType, and improvementNotation testing

### LINKED ISSUE #580: Testing enhancements measure scoring, improvementNotation, reportType
# Description 

Use #567 introduced all encompassing cql and test data generator for unit testing additional areas of quality measures. The eventual goal being that Measure Evaluation processes and logic can have basic resources used to validate specific scenarios are performing correctly, instead of using complex scenarios that are difficult to root cause issues in.

This is found in SimpleCQL.cql

The overall design allows for one Library and CQL to be used for constructing any MeasureScoringType and a TestDataGeneration class to create patient populations of varying sizes.

The scope of this change is aimed at these areas:

Measure Improvement Notation
Measure Scoring Type testing
Measure EvalType/ReportType testing

This will remove redundant test classes and associated test data (if not used by other tests):

CohortMeasureProcessorTest.java
ContinuousVariableMeasureProcessorTest.java
EncounterBasisMeasureProcessorTest.java
MinimalMeasureEvaluationTest.java
R4MeasureServiceTest.java

This will add test classes and measure scenario test data around

Measure Improvement Notation
Measure Scoring Type testing
Measure EvalType/ReportType testing

Creating New Test classes aimed at focused coverage for the given scenario:

MeasureReportTypeIndividual
MeasureReportTypeSummary
MeasureReportTypeSubjectList
MeasureScoringTypeCohort
MeasureScoringTypeContinuousVariable
MeasureScoringTypeProportion
MeasureScoringTypeRatio
MeasureImprovementNotationTest

# Enhancements

- R4EvalType - enum allowing for definition of accepted values to be enforced (previously accepted parameter values that are not to spec (r4 patient-list, patient)
- MeasureScoringTypePopulations - for defining allowed and required populations given a scoring type, allowing for one place to manage acceptable combinations.
- Additional Testing methods to expand coverage aimed at scoringtypes, improvementNotation, and evalTypes
- fixed a bug on setting groupDef or measureDef improvement notation that was allowing wrong value to be set

### PR DESCRIPTION
**Pull request by Capt-Mac**
closes #580

### CODE REVIEW

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/common/MeasureEvaluator.java [hunk 1]
```diff
@@ -410,6 +412,11 @@ protected void evaluateProportion(
             int populationSize,
             MeasureReportType reportType,
             EvaluationResult evaluationResult) {
+        // check populations
+        R4MeasureScoringTypePopulations.validateScoringTypePopulations(
+                groupDef.populations().stream().map(PopulationDef::type).collect(Collectors.toList()),
+                MeasureScoring.fromCode(groupDef.measureScoring().toCode()));
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/common/MeasureEvaluator.java:418]: Why do this?  Why not just pass groupDef.measureScoring() directly into the method?

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureDefBuilder.java [hunk 2]
```diff
@@ -167,25 +168,51 @@ private MeasureScoring getMeasureScoring(Measure measure) {
...
+                || IMPROVEMENT_NOTATION_SYSTEM_DECREASE.equals(improvementNotationValue);
+        if (!hasValidCode) {
+            throw new IllegalArgumentException(String.format(
+                    "ImprovementNotation Coding has invalid code: %s, combination for Measure.",
+                    improvementNotationValue));
+        }
+    }
+
+    public String getGroupImprovementNotationExt(Measure.MeasureGroupComponent group) {
+        StringType value = (StringType) group.getExtensionByUrl(MEASUREREPORT_IMPROVEMENT_NOTATION_EXTENSION)
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureDefBuilder.java:186]: Wll this always be a StringType?  Would it be a good idea to do an instanceof check here first?

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureDefBuilder.java [hunk 3]
```diff
@@ -167,25 +168,51 @@ private MeasureScoring getMeasureScoring(Measure measure) {
...
+        boolean hasValidCode = IMPROVEMENT_NOTATION_SYSTEM_INCREASE.equals(improvementNotationValue)
+                || IMPROVEMENT_NOTATION_SYSTEM_DECREASE.equals(improvementNotationValue);
+        if (!hasValidCode) {
+            throw new IllegalArgumentException(String.format(
+                    "ImprovementNotation Coding has invalid code: %s, combination for Measure.",
+                    improvementNotationValue));
+        }
+    }
+
+    public String getGroupImprovementNotationExt(Measure.MeasureGroupComponent group) {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/dstu3/Dstu3MeasureDefBuilder.java:185]: Do any of these last 3 methods really need to be public?

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java [hunk 4]
```diff
@@ -149,7 +148,19 @@ && checkPopulationForCode(populations, DATEOFCOMPLIANCE) == null) {
                 measure.getVersion(),
                 groups,
                 sdes,
-                measureBasisDef.isBooleanBasis(measure));
+                measureBasisDef.isBooleanBasis(measure),
+                useMeasureImprovementNotation(groups));
+    }
+
+    public boolean useMeasureImprovementNotation(List<GroupDef> groups) {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java:155]: Does this need to be public?

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java [hunk 5]
```diff
@@ -223,25 +244,43 @@ private MeasureScoring getGroupMeasureScoring(MeasureScoring measureLevelScoring
...
     }
 
-    public boolean groupIsIncreaseImprovementNotation(
-            boolean measureImprovementNotationIsIncrease, MeasureGroupComponent group) {
-        var improvementNotationExt = group.getExtensionByUrl(MEASUREREPORT_IMPROVEMENT_NOTATION_EXTENSION);
-        if (improvementNotationExt != null) {
-            var code = (CodeableConcept) improvementNotationExt.getValue();
-            return isIncreaseImprovementNotation(code);
+    public CodeableConcept getGroupImprovementNotationExt(MeasureGroupComponent group) {
+        return (CodeableConcept) group.getExtensionByUrl(MEASUREREPORT_IMPROVEMENT_NOTATION_EXTENSION)
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureDefBuilder.java:266]: Do you need an instanceof check here?

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureEvalType.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
+package org.opencds.cqf.fhir.cr.measure.r4;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureEvalType.java:6]: Question:  Do we not generate copyright headers for our source files?

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureScoringTypePopulations.java [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        public MeasurePopulationType getPopulationType() {
+            return measurePopulationType;
+        }
+
+        public static Set<MeasurePopulationType> getPopulations() {
+            EnumSet<R4MeasureScoringTypePopulations.PROPORTION_ALLOWED> data = EnumSet.allOf(PROPORTION_ALLOWED.class);
+            return data.stream().map(PROPORTION_ALLOWED::getPopulationType).collect(Collectors.toSet());
+        }
+
+        public static void isMember(List<MeasurePopulationType> populations) {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureScoringTypePopulations.java:37]: Suggest you rename this method to "validateMember()".  "is" implies you're returning a boolean.

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureScoringTypePopulations.java [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+            return measurePopulationType;
+        }
+
+        public static Set<MeasurePopulationType> getPopulations() {
+            EnumSet<R4MeasureScoringTypePopulations.PROPORTION_REQUIRED> data =
+                    EnumSet.allOf(PROPORTION_REQUIRED.class);
+            return data.stream().map(PROPORTION_REQUIRED::getPopulationType).collect(Collectors.toSet());
+        }
+
+        public static void hasRequired(List<MeasurePopulationType> populations) {
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureScoringTypePopulations.java:69]: Same thing here:  "has" implies a boolean.

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/Measure.java [hunk 9]
```diff
@@ -506,6 +544,35 @@ private List<String> getContainedIdsPerResourceType(ResourceType theResourceType
...
+        /**
+         * Subject-List will contain Lists of resource references to represent population. This test validates that correct resourceType is present.
+         * TODO: if group specifies basis instead of Measure then this will need to be updated.
+         * @param resourceType resource basis of population (encounter or patient)
+         * @return value allowing chaining of more methods
+         */
+        public SelectedReport subjectResultsHaveResourceType(String resourceType) {
+            var lists = value().getContained().stream()
+                    .filter(t -> t.getResourceType().equals(ResourceType.List))
+                    .map(x -> (ListResource) x)
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/Measure.java:556]: Please add a filter for instanceof ListResource

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureImprovementNotationTest.java [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+    // group defined
+    // increase
+    // decrease
+    // invalid value
+    // multi-group
+    private static final String CLASS_PATH = "org/opencds/cqf/fhir/cr/measure/r4";
+    private static final Repository repository = new IgRepository(
+            FhirContext.forR4Cached(),
+            Paths.get(getResourcePath(MeasureImprovementNotationTest.class) + "/" + CLASS_PATH + "/" + "MeasureTest"));
+    protected Given given = Measure.given().repository(repository);
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureImprovementNotationTest.java:30]: If you're not planning to mutate this field, please make it final.

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureReportTypeIndividualTest.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+import org.opencds.cqf.fhir.cr.measure.r4.utils.TestDataGenerator;
+import org.opencds.cqf.fhir.utility.repository.ig.IgRepository;
+
+/**
+ * This test is to verify and confirm that created Individual report has required format and output
+ *      subject field has reference
+ *      report has date
+ *      report type summary
+ *      has measure url
+ *      has correct status
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureReportTypeIndividualTest.java:23]: I see only 2 tests in this class.  Is the point to capture this list of test cases in a future PR?

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureReportTypeSummaryTest.java [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+
+/**
+ * This test is to verify and confirm that created Summary report has required format and output
+ *     subject field has reference
+ *     report has date
+ *     report type summary
+ *     has measure url
+ *     has correct status
+ *     has period
+ *     errors when missing required population
```
Comment (lukedegruchy) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureReportTypeSummaryTest.java:25]: There are no error tests in this test class either

**Review by lukedegruchy — APPROVED** [review]

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureEvalType.java (hunk 6 above)
Comment (Capt-Mac) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/measure/r4/R4MeasureEvalType.java:6]: we do not

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureReportTypeIndividualTest.java (hunk 11 above)
Comment (Capt-Mac) [reply cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureReportTypeIndividualTest.java:23]: these tests capture correct type and meta data. scoring type should be checking for correct populations, fixed description

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureReportTypeSummaryTest.java (hunk 12 above)
Comment (Capt-Mac) [reply cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/measure/r4/MeasureReportTypeSummaryTest.java:25]: fixed description

**Pull request merged**